internals/cli/masker/writer.go

```diff
@@ -0,0 +1,204 @@
+			return len(m.sequence)
+		}
+		return 0
+	}
+
+	m.findShift()
+
+	if m.sequence[m.currentIndex] == in {
+		return m.Read(in)
+	}
+	m.currentIndex = 0
+	return 0
+}
+
+// findShift checks whether we can also make a partial match by decreasing the currentIndex .
+// For example, if the sequence is foofoobar, if someone inserts foofoofoobar, we still want to match.
+// So after the third f is inserted, the currentIndex is decreased by 3 with the following code.
+func (m *sequenceMatcher) findShift() {
+	for offset := 1; offset <= m.currentIndex; offset++ {
+		ok := true
+		for i := 0; i < m.currentIndex-offset; i++ {
+			if m.sequence[i] != m.sequence[i+offset] {
+				ok = false
+				break
+			}
+		}
+		if ok {
+			m.currentIndex -= offset
+			return
+		}
+	}
+}
+
+// InProgress returns whether this sequenceMatcher is currently partially matching.
+//
+// For example, if the sequence is "foobar" and the registered input is "foob", InProgress() returns true.
+func (m *sequenceMatcher) InProgress() bool {
+	return m.currentIndex > 0
+}
+
+// Reset forgets the current match.
+func (m *sequenceMatcher) Reset() {
+	m.currentIndex = 0
+}
+
+// maskByte represents a byte and whether the byte should be masked or not.
+type maskByte struct {
+	byte
+	masked bool
+}
+
+// MaskedWriter wraps an io.Writer and masks all matches found by Matchers with MaskString.
+// If no write is made for Timeout on the io.Writer, any matches in progress are reset
+// and the buffer is flushed. This is to ensure that the writer does not hang on partial matches.
+type MaskedWriter struct {
+	w          io.Writer
+	MaskString string
+	Matchers   []Matcher
+	Timeout    time.Duration
+
+	buf    []maskByte
+	lock   *sync.Mutex
+	output chan []maskByte
+	err    error
+	nIn    int64
+	nOut   int64
+}
+
+func NewMaskedWriter(w io.Writer, masks [][]byte, maskString string, timeout time.Duration) *MaskedWriter {
+	var lock sync.Mutex
+	matchers := make([]Matcher, len(masks))
+	for i, mask := range masks {
+		matchers[i] = &sequenceMatcher{
+			sequence: mask,
+		}
+	}
+	return &MaskedWriter{
+		w:          w,
+		MaskString: maskString,
+		Matchers:   matchers,
+		Timeout:    timeout,
+		lock:       &lock,
+		output:     make(chan []maskByte, 1),
+	}
+}
+
+// Write implements implements to Write from io.Writer
+// It is responsible for finding any matches to mask and mark the appropriate bytes as masked.
+// This function never returns an error. These can instead be caught with Flush().
+func (mw *MaskedWriter) Write(p []byte) (n int, err error) {
+	for _, b := range p {
+		matchInProgress := false
+
+		mw.lock.Lock()
+		mw.buf = append(mw.buf, maskByte{byte: b})
+
+		for _, matcher := range mw.Matchers {
+			maskLen := matcher.Read(b)
+			for i := 0; i < maskLen; i++ {
+				mw.buf[len(mw.buf)-1-i].masked = true
+			}
+			matchInProgress = matchInProgress || matcher.InProgress()
+		}
+
+		if !matchInProgress {
+			mw.flushBuffer()
+		} else {
+			mw.output <- []maskByte{}
```

Comment:
What's this for?

Reply:
To reset the timeout. This way every call to `Write` resets the timeout. But thinking of it, it might make more sense not to do this. Then the timeout actually defines the maximum time between two flushes.

Reply:
Changed it.